flatbuffers/flatbuffers_case_study.md

```diff
@@ -0,0 +1,151 @@
+
+All this data must be stored and accessible for the recorders counterpart - the player. The player knows how to take the mass of events and reconstruct them into a watchable clip.
+
+This undertaking raises many challenges, one of which is to transfer all of those events from the recorded user to our servers, and from our servers to the player. The efficiency of this process is measured both by the size of the transfered data and by the time it takes to parse it, measures that JSON does not excel in.
+
+# What are FlatBuffers?
+[FlatBuffers](https://google.github.io/flatbuffers) is a serialization library by Google. It was first made by Google with android games in mind, but now it has ports to several languages including JavaScript.
+
+The main goal of FlatBuffers is to avoid deseriazation what so ever. This is achieved by defining a binary data protocol - a well defined method to convert data to, and from binary. The binary structure created by this protocol can be sent over the wire, and read without further processing. To reiterate, when transferring JSON, we need to transform data into a string, send the string over the wire, parse the string, and transform it into a native object. Flatbuffers does not require any of it. You start up with binary, send the same binary, and read directly from binary.
+
+In addition to the parsing efficiency the binary format gains us another advantage - binary representation of data is often more size efficient. We can store a 10 digit int using 4 UInt bytes, instead of 10 chars.
+
+## Flat buffers Example
+This won't be a FlatBuffers tutorial, but to give you a feel for what it's like lets took on a simplified version of our data model.
+
+The first step of FlatBuffers is to define an *idl* file. This file defines the scheme of our data. We define an ```Event``` object that stores event_type, timestamp, and event_data. Each event type has it's own data object
+
+```idl
+// we can define enums
+enum EventType:short {
+  MOUSE_CLICK,
+  MOUSE_POSITION,
+  INPUT,
+  DOM_MUTATION
+}
+
+// a table is similar to struct definition
+table MouseClickData{
+  x: ushort;
+  y: ushort;
+  target_node_id: ushort;
+}
+
+table MousePositionData{ /* ... */ }
+table InputData{ /* ... */ }
+table DomMutationData{ /* ... */ }
+
+// union allows us to store different types of data in the same field
+union EventData {
+  MouseClickData,
+  MousePositionData,
+  InputData,
+  DomMutationData
+}
+
+table Event {
+  timestamp: ulong;
+  type: EventType;
+  event_data: EventData;
+}
+
+root_type Event;
+```
+
+After defining the schema, we run the flatbuffer 'compiler' - flatc. This generates code in our language of choice. That code handles all the bytes and bits calculation flatbuffer is based on.
```

Comment:
one thing to mention is that flatbuffer are available to a LOT of languages aside from js